concordium-consensus/src/Concordium/GlobalState/LMDB/Helpers.hs

```diff
@@ -28,6 +30,158 @@ import Foreign.Marshal.Utils
 import Foreign.Marshal.Alloc
 import Foreign.Ptr
 import Foreign.Storable
+import Lens.Micro.Platform
+
+
+-- Implementation of reader-writer locks that strongly prefer writers.
+
+data RWState =
+   -- |Nobody has acquired the lock.
+  Free { waitingWriters :: !Word64 }
+  -- |The given number of readers have acquired the lock.
+  | ReadLocked {
+      readers :: !Word64,
+      waitingWriters :: !Word64
+      }
+  -- |The lock is acquired by a single writer.
+  | WriteLocked {
+      waitingWriters :: !Word64
+      }
+    deriving(Show)
+
+-- |A reader writer lock with a bias for writers. If there are waiting writers
+-- readers will block.
+data RWLock = RWLock {
+  state :: !(MVar RWState),
+  -- |Used to signal waiting readers to wake up. This is empty when there is at
+  -- least one active reader and full otherwise.
+  readLock :: !(MVar ()),
+  -- |Used to signal waiting writers to wake up. This is empty when there is an
+  -- active writer, and full otherwise.
+  writeLock :: !(MVar ())
+  }
+
+-- |Initialize a lock in the unlocked state.
+initializeLock :: IO RWLock
+initializeLock = do
+  state <- newMVar (Free 0)
+  readLock <- newMVar ()
+  writeLock <- newMVar ()
+  return RWLock{..}
+
+-- |Acquire a read lock. This will block until there are no pending writers
+-- waiting to acquire the lock.
+acquireRead :: RWLock -> IO ()
+acquireRead RWLock{..} = mask_ go
+  where
+    go = takeMVar state >>= \case
+      st@(Free waitingWriters)
+        | waitingWriters == 0 -> do
+            -- the lock is free and there are no waiting writers. Acquire a read lock.
+            takeMVar readLock
+            putMVar state (ReadLocked 1 0)
+        | otherwise -> do
+            -- the lock is free, but there are waiting writers. We do nothing and try again.
+            -- Due to fairness of MVars next time another thread will make progress
+            -- so we are going to end up after a finite number of iterations, in a WriteLocked state.
+            putMVar state st
+            go
+      st@(ReadLocked n waitingWriters)
+          | waitingWriters == 0 ->
+            -- No waiting writers, add another reader.
+            putMVar state $! ReadLocked (n + 1) 0
+          | otherwise -> do
+              -- Some readers hold the lock, but there are waiting writers.
+              -- We do nothing and wait until there are no more readers and attempt again.
+              -- At that point we are likely to end up in WriteLocked state.
+              putMVar state st
+              readMVar readLock
+              go
+      lockState@(WriteLocked _) -> do
+        -- There is an active writer. Do nothing and wait until the writer is done.
+        putMVar state lockState
+        readMVar writeLock
+        go
+
+
+acquireWrite :: RWLock -> IO ()
+acquireWrite RWLock{..} = mask_ $ go False
+  where
+    -- the boolean flag indicates whether this is a first iteration of the loop (False) or not (True)
```

Comment:
Maybe add a remark of what the purpose of the boolean flag. Something along the lines of "... waitingWriters field is updated according to the the boolean flag i.e. the waitingWriters field is incremented if a new writer has arrived... " - Now you're at it (adding a comment)